src/main/osd/osd_elements.c

```diff
@@ -2094,81 +2168,131 @@ uint8_t osdGetActiveElementCount(void)
         // If the background layer isn't supported then we
         // have to draw the element's static layer as well.
-        osdDrawSingleElementBackground(osdDisplayPort, activeOsdElementArray[activeElement]);
-    }
+        backgroundRendered = osdDrawSingleElementBackground(osdDisplayPort, activeOsdElementArray[activeElement]);
 
-    osdDrawSingleElement(osdDisplayPort, activeOsdElementArray[activeElement]);
+        return retval;
+    }
 
-    if (++activeElement >= activeOsdElementCount) {
-        activeElement = 0;
-        retval = false;
+    // Only advance to the next element if rendering is complete
+    if (osdDrawSingleElement(osdDisplayPort, activeOsdElementArray[activeElement])) {
+        // Prepare to render the background of the next element
+        backgroundRendered = false;
+        if (++activeElement >= activeOsdElementCount) {
+            activeElement = 0;
+            retval = false;
+        }
     }
 
     return retval;
 }
 
 #ifdef USE_SPEC_PREARM_SCREEN
-void osdDrawSpec(displayPort_t *osdDisplayPort)
+bool osdDrawSpec(displayPort_t *osdDisplayPort)
 {
-    if (!ARMING_FLAG(ARMED) && osdConfig()->osd_show_spec_prearm) {
-        const uint8_t midRow = osdDisplayPort->rows / 2;
-        const uint8_t midCol = osdDisplayPort->cols / 2;
+    static enum {CLR, RPM, POLES, MIXER, THR, MOTOR, BAT, VER} specState = CLR;
+    static int currentRow;
 
-        char buff[OSD_ELEMENT_BUFFER_LENGTH] = "";
+    const uint8_t midRow = osdDisplayPort->rows / 2;
+    const uint8_t midCol = osdDisplayPort->cols / 2;
 
-        memset(buff,0,strlen(buff));
-        int len = 0;
-        int currentRow = midRow - 3;
+    char buff[OSD_ELEMENT_BUFFER_LENGTH] = "";
+
+    memset(buff, 0, strlen(buff));
+    int len = 0;
+
+    switch (specState) {
+    default:
+    case CLR:
+        displayClearScreen(osdDisplayPort, DISPLAY_CLEAR_NONE);
+        currentRow = midRow - 3;
+
+        specState = RPM;
+        break;
 
+    case RPM:
 #ifdef USE_RPM_LIMIT
-        const bool rpmLimitActive = mixerConfig()->rpm_limit > 0 && isMotorProtocolBidirDshot();
-        if (rpmLimitActive) {
-            len = tfp_sprintf(buff, "RPM LIMIT ON  %d", mixerConfig()->rpm_limit_value);
-        } else {
-            len = tfp_sprintf(buff, "%s", "RPM LIMIT OFF");
+        {
+            const bool rpmLimitActive = mixerConfig()->rpm_limit > 0 && isMotorProtocolBidirDshot();
+            if (rpmLimitActive) {
+                len = tfp_sprintf(buff, "RPM LIMIT ON  %d", mixerConfig()->rpm_limit_value);
+            } else {
+                len = tfp_sprintf(buff, "%s", "RPM LIMIT OFF");
+            }
+            displayWrite(osdDisplayPort, midCol - len/2, currentRow++, DISPLAYPORT_SEVERITY_NORMAL, buff);
+
+            if (rpmLimitActive) {
+                specState = POLES;
+            } else {
+                specState = THR;
+            }
         }
+        break;
+
+    case POLES:
+        memset(buff,0,strlen(buff));
```

Comment:
(this is nop, see above. Also tfp_sprintf overwrites buff and zero-terminates it.,

Reply:
Done

Reply:
Done